modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java

```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
+      //
+      // The sorting is, again, not for correctness, because consumers of this API need to be
+      // able to deal with that. However, sorting this here will result in fewer temporary objects
+      // or invalid states in the consumer.
+      final ArrayList<Item> items = Stream.concat(eventItems, seriesItems)
           .collect(Collectors.toCollection(ArrayList::new));
+      items.sort(Comparator.comparing(item -> item.getModifiedDate()));
+
+
+      // Obtain information to allow Tobira to plan the next harvesting request.
+      boolean hasMore = hasMoreEvents || hasMoreSeriesInRange;
+      final Date includesItemsUntilRaw;
+      if (!hasMoreEvents && !hasMoreSeriesInRange) {
+        // All events and series up to now have been harvested.
+        includesItemsUntilRaw = new Date();
+      } else if (!hasMoreEvents && hasMoreSeriesInRange) {
+        // `rawEvents` contains all events that currently exist. Our response won't contain the ones
+        // that have a modified date after the one of the last raw series. But that means that we
+        // will return all series and events until the last raw series.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      } else if (hasMoreEvents && !hasMoreSeriesInRange) {
+        // There are more events, but no additional series in the range from `since` to the modified
+        // date of the last raw event. So we know there are no other events or series before the
+        // last raw events.
+        includesItemsUntilRaw = rawEvents[rawEvents.length - 1].getModified();
+      } else {
+        // There are more events and more series in the given range. In theory, this would be
+        // `Math.min()` of the last raw event and last raw series. However, since `hasMoreEvents`
+        // is set, we only loaded series with modified dates smaller than the last raw event's
+        // modified date. That means, that the modified date of the last raw series is always
+        // smaller than that of the last raw event.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      }
+
+      // The `includesItemsUntil` we return has to be at least `TIME_BUFFER_SIZE` in the past. See
+      // the constant's documentation for more information on that.
+      final long includesItemsUntil = Math.min(
+          includesItemsUntilRaw.getTime(),
+          new Date().getTime() - TIME_BUFFER_SIZE
+      );
 
-      // Assembly full response
+
+      // Assembly full response.
+      final List<Jsons.Val> outItems = items.stream()
+          .map(item -> item.getJson())
+          .collect(Collectors.toCollection(ArrayList::new));
       final Jsons.Obj json = Jsons.obj(
           Jsons.p("includesItemsUntil", includesItemsUntil),
           Jsons.p("hasMore", hasMore),
-          Jsons.p("items", Jsons.arr(items))
+          Jsons.p("items", Jsons.arr(outItems))
```

Comment:
_(Applied to all JSON related code)_

It's certainly personal preference, but I find it much easier to work with GSON instead of json-simple since you don't need to manually build all your JSON objects, which usually also makes the code much easier to read:

```
Series series = new Series();
…
String jsonString = gson.toJson(series);
```

or

```
List<Series> seriesList = new ArrayList<>();
…
String jsonString = gson.toJson(seriesList);
```

Of course, you can still build it manually, if you need to:
```
Map<String, Object> customData = new Map<>();
customData.put("key", "value");
String jsonString = gson.toJson(customData);
```

Reply:
I am generally also a fan of these kinds of serialization libraries. On the Tobira side, we do it that way. But in this case I'm not sure if it would really buy us anything. As explained in private chat: to have complete control over the format of the API (which we should, IMO) we can't call `toJson` on the objects returned by the search and series API. Because those classes can be modified at any time, potentially breaking API compatibility. Thus, we need our own classes. But defining them, including fields, getters, setter AND the code to create them would add a lot of code to this module. So in this case, I think the json-simple works fine and results in less code. So I'd like to keep it for now.